### PR TITLE
Object state manager for 2.x

### PR DESCRIPTION
**Pull request by phalasz**
Ability to manage `ObjectState`'s and `ObjectStateGroup`'s.

### CODE REVIEW

File: API/Collection/ObjectStateCollection.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+namespace Kaliop\eZMigrationBundle\API\Collection;
+
+use eZ\Publish\API\Repository\Values\ObjectState\ObjectState;
+
+/**
+ * @todo add phpdoc to suggest typehinting
+ */
+class ObjectStateCollection extends AbstractCollection
+{
+    protected $allowedClass = ObjectState::class;
```
Comment (gggeek) [inline API/Collection/ObjectStateCollection.php]: Cool, I never thought about his syntax. Is it compatible all the way back to old pp versions ?

File: Core/Executor/ObjectStateManager.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        foreach(array('object_state_group_id', 'names', 'identifier') as $key) {
+            if (!isset($this->dsl[$key])) {
+                throw new \Exception("The '$key' key is missing in a object state creation definition");
+            }
+        }
+
+        $objectStateService = $this->repository->getObjectStateService();
+
+        $objectStateGroupId = $this->dsl['object_state_group_id'];
+        if ($this->referenceResolver->isReference($objectStateGroupId)) {
```
Comment (gggeek) [inline Core/Executor/ObjectStateManager.php]: nitpick: we now have a single-function call in the API for this

File: Core/Executor/ObjectStateManager.php [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            }
+        }
+
+        if (0 === count($objectStateCreateStruct->names)) {
+            throw new \Exception('No object state names have been defined. Need to specify at least one to create the state.');
+        }
+
+        $objectState = $objectStateService->createObjectState($objectStateGroup, $objectStateCreateStruct);
+
+        $this->setReferences($objectState);
```
Comment (gggeek) [inline Core/Executor/ObjectStateManager.php]: you need to return the object state, for event listeners

File: Core/Executor/ObjectStateManager.php [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+            foreach ($this->dsl['names'] as $name) {
+                $objectStateUpdateStruct->names[$name['languageCode']] = $name['name'];
+            }
+
+            $state = $objectStateService->updateObjectState($state, $objectStateUpdateStruct);
+
+            $this->setReferences($state);
+        }
+
```
Comment (gggeek) [inline Core/Executor/ObjectStateManager.php]: you need to return the object state collection, for event listeners

**Comment by gggeek** [comment]
PS: sorry, I just realized that the PR is against branch 2.x, whereas the comments I added where about about branch 3. Sorry about that

File: API/Collection/ObjectStateCollection.php (hunk 1 above)
Comment (phalasz) [inline API/Collection/ObjectStateCollection.php]: According to Andre you need atleast PHP 5.5

**Comment by gggeek** [comment]
PS: I am refactoring this PR a bit.
Apart from extra capabilities, the main difference is that I will only support the 'match' syntax to identify states and state-groups in the DSL for update and deletion, as that is the only supported one going forward, for all types of entities.

**Pull request merged**